### ISSUE TITLE
Overridable bash template

### ISSUE THREAD

**Issue by damirv** [sbt/sbt-native-packager#635]
I'm using the Play framework and I want to override the bash template. Is there currently a way to override the bash template without placing it under `src/templates/bash-template`? 

It seems from [this](https://github.com/sbt/sbt-native-packager/blob/master/src/main/scala/com/typesafe/sbt/packager/archetypes/JavaApp.scala#L105) that it uses `sourceDirectory` to determine the base. I tried setting `sourceDirectory in makeBashScript := baseDirectory.value / "app"` and that doesn't work (as I expected). Play defines the following: `sourceDirectory in Compile := baseDirectory.value / "app",`. Setting `sourceDirectory := baseDirectory.value / "app"` works but then I get also other folders in the package.

**Comment by muuki88** [comment]
I'm afraid this is not possible without overriding the task that generates the bash script.
We could introduce a special setting, but I think this more cosmetic than actual functionality.

**Comment by damirv** [comment]
With overriding the task, you mean something like [this](https://github.com/sbt/sbt-native-packager/blob/master/src/main/scala/com/typesafe/sbt/packager/archetypes/JavaApp.scala#L262)?

**Comment by damirv** [comment]
I have this now (added to build.sbt) and it works (bash-template is located in the `conf` folder of Play). But if we would have `defaultTemplateLocation` as a setting, then it would be easier to override (because that's basically what I'm overriding in this function). Do you think this is useful?

```scala
{
  object Override extends JavaAppStartScript {
    val bashTemplate = "bash-template"

    def settings: Seq[Setting[_]] = Seq(
      makeBashScript <<= (baseDirectory, bashScriptDefines, target in Universal, executableScriptName, sourceDirectory) map makeCustomUniversalBinScript
    )

    import JavaAppPackaging.autoImport._

    def makeCustomUniversalBinScript(baseDir: File, defines: Seq[String], tmpDir: File, name: String, sourceDir: File): Option[File] =
      if (defines.isEmpty) None
      else {
        val defaultTemplateLocation = baseDir / "conf" / bashTemplate
        val defaultTemplateSource = getClass getResource bashTemplate

        val template = if (defaultTemplateLocation.exists)
          defaultTemplateLocation.toURI.toURL
        else defaultTemplateSource

        val scriptBits = JavaAppBashScript.generateScript(defines, template)
        val script = tmpDir / "tmp" / "bin" / name
        IO.write(script, scriptBits)
        // TODO - Better control over this!
        script.setExecutable(true)
        Some(script)
      }
  }
  Override.settings
}
```

**Comment by muuki88** [comment]
Hm. I'm not sure if this setting really makes sense as it's more a cosmetic thing. Would creating a src folder have cause any issues?

On the other hand, separating is not that of a big deal.

**Comment by muuki88** [comment]
And thanks for sharing your code snippet. A few notes/questions

- why check if defines isEmpty?
- getClass getResource won't work I think

**Comment by damirv** [comment]
If it's not a lot of hassle, I'd would rather add it, as it's nice to stick to the directory convention of the project you're using (in my case, Play, which doesn't have a src folder).

I am not sure why the check for `defines.isEmpty` exists, I copied that method straight from the source code (except for the line `val defaultTemplateLocation = baseDir / "conf" / bashTemplate`). Same goes for `getClass getResource bashTemplate`. The original method can be found [here](https://github.com/sbt/sbt-native-packager/blob/master/src/main/scala/com/typesafe/sbt/packager/archetypes/JavaApp.scala#L284).

**Comment by muuki88** [comment]
You can change the play layout structure since 2.4 if that helps. I keep this issue then as a feature request. Happy to merge a pr from you ;)

> I am not sure why the check for defines.isEmpty exists, I copied that method straight from the source code 

That's weird. I have to look closer.

**Comment by muuki88** [comment]
Implemented in #646
